server/src/main/java/io/druid/segment/realtime/RealtimeManager.java

```diff
@@ -294,60 +265,69 @@ public FireDepartmentMetrics getMetrics()
-          runFirehoseV2(firehoseV2);
-        } else {
-          firehose = initFirehose();
-          runFirehose(firehose);
-        }
+        final Closer closer = Closer.create();
 
+        try {
+          Object metadata = plumber.startJob();
+
+          Firehose firehose;
+          FirehoseV2 firehoseV2;
+          final boolean success;
+          if (fireDepartment.checkFirehoseV2()) {
+            firehoseV2 = initFirehoseV2(metadata);
+            closer.register(firehoseV2);
+            success = runFirehoseV2(firehoseV2);
+          } else {
+            firehose = initFirehose();
+            closer.register(firehose);
+            success = runFirehose(firehose);
+          }
+          if (success) {
+            // pluber.finishJob() is called only when every processing is successfully finished.
+            closer.register(() -> plumber.finishJob());
+          }
+        }
+        catch (InterruptedException e) {
+          log.warn("Interrupted while running a firehose");
+          throw closer.rethrow(e);
+        }
+        catch (Exception e) {
+          log.makeAlert(
+              e,
+              "[%s] aborted realtime processing[%s]",
+              e.getClass().getSimpleName(),
+              fireDepartment.getDataSchema().getDataSource()
+          ).emit();
+          throw closer.rethrow(e);
+        }
+        catch (Error e) {
+          log.makeAlert(e, "Error aborted realtime processing[%s]", fireDepartment.getDataSchema().getDataSource())
+             .emit();
+          throw closer.rethrow(e);
+        }
+        finally {
+          closer.close();
+        }
       }
-      catch (RuntimeException e) {
-        log.makeAlert(
-            e,
-            "[%s] aborted realtime processing[%s]",
-            e.getClass().getSimpleName(),
-            fireDepartment.getDataSchema().getDataSource()
-        ).emit();
-        normalExit = false;
+      catch (IOException e) {
         throw Throwables.propagate(e);
       }
-      catch (Error e) {
-        log.makeAlert(e, "Error aborted realtime processing[%s]", fireDepartment.getDataSchema().getDataSource())
-           .emit();
-        normalExit = false;
-        throw e;
-      }
-      finally {
-        CloseQuietly.close(firehose);
-        if (normalExit) {
-          plumber.finishJob();
-          plumber = null;
-          firehose = null;
-        }
-      }
     }
 
-    private void runFirehoseV2(FirehoseV2 firehose)
+    private boolean runFirehoseV2(FirehoseV2 firehose) throws Exception
     {
-      try {
-        firehose.start();
-      }
-      catch (Exception e) {
-        log.error(e, "Failed to start firehoseV2");
-        return;
-      }
+      firehose.start();
 
       log.info("FirehoseV2 started");
       final Supplier<Committer> committerSupplier = Committers.supplierFromFirehoseV2(firehose);
       boolean haveRow = true;
       while (haveRow) {
+        if (Thread.interrupted()) {
```

Comment:
It doesn't seem very reliable to me, because sometimes interruption status may be consumed (due to bugs). So in addition suggested to add "stopping" flag to `RealtimeManager` and check it here in addition to interruption status.